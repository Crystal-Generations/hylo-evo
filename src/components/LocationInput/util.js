export function convertMapboxToLocation (mapboxResult) {
  const neighborhoodObject = mapboxResult.context.find(c => c.id.includes('neighborhood'))
  const postcodeObject = mapboxResult.context.find(c => c.id.includes('postcode'))
  const placeObject = mapboxResult.context.find(c => c.id.includes('place'))
  const regionObject = mapboxResult.context.find(c => c.id.includes('region'))
  const countryObject = mapboxResult.context.find(c => c.id.includes('country'))

  return {
    accuracy: mapboxResult.properties.accuracy,
    addressNumber: mapboxResult.address,
    addressStreet: mapboxResult.text,
    bbox: mapboxResult.bbox ? [{ lng: mapboxResult.bbox[0], lat: mapboxResult.bbox[1] }, { lng: mapboxResult.bbox[2], lat: mapboxResult.bbox[3] }] : [],
    center: { lng: mapboxResult.center[0], lat: mapboxResult.center[1] },
    city: placeObject && placeObject.text,
    country: countryObject && countryObject.short_code,
    fullText: mapboxResult.place_name,
    // geometry: [Point]
    // locality
    neighborhood: neighborhoodObject && neighborhoodObject.text,
    region: regionObject && regionObject.text,
    postcode: postcodeObject && postcodeObject.text
    // wikidata: String
  }
}
