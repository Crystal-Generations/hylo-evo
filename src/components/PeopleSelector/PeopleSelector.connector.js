import { connect } from 'react-redux'

import { addParticipant, removeParticipant, fetchPeople, matchesSelector, participantsSelector, setAutocomplete } from './PeopleSelector.store'

export function mapStateToProps (state) {
  return {
    autocomplete: state.autocomplete,
    matches: matchesSelector(state),
    participants: participantsSelector(state)
  }
}

export default connect(mapStateToProps, { addParticipant, removeParticipant, fetchPeople, setAutocomplete })
