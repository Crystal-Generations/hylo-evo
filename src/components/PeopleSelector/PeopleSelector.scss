.people-selector {
  flex-grow: 2;
}

.autocomplete-control {
  display: flex;
  flex-grow: 2;
  border-bottom: 1px $color-rhino-30 solid;
  padding-bottom: $space-2x;
}

.autocomplete {
  flex-grow: 2;
  border: none;
  outline: none;

  // Yep, believe it or not this is necessary... can't group the selectors!
  &::-webkit-input-placeholder {
    @include font-regular;
    color: $color-rhino-30;
    font-size: 14px;
    letter-spacing: 0.4;
    line-height: 18px;
  }

  &::-moz-placeholder {
    @include font-regular;
    color: $color-rhino-30;
    font-size: 14px;
    letter-spacing: 0.4;
    line-height: 18px;
  }

  &:-ms-input-placeholder {
    @include font-regular;
    color: $color-rhino-30;
    font-size: 14px;
    letter-spacing: 0.4;
    line-height: 18px;
  }

  &:-moz-placeholder {
    @include font-regular;
    color: $color-rhino-30;
    font-size: 14px;
    letter-spacing: 0.4;
    line-height: 18px;
  }

  &::placeholder {
    @include font-regular;
    color: $color-rhino-30;
    font-size: 16px;
    letter-spacing: 0.4;
    line-height: 20px;
  }
}


.close-button {
  margin-bottom: $space-4x;
  margin-left: $space-6x;
}

.thread-header {
  display: flex;
  flex-grow: 2;
  align-items: flex-end;
  height: $space-20x;
  border: $light-border;
  padding: $space-3x $space-7x;
  padding-top: 0;

  &:focus, &:hover {
    outline: none;
    -webkit-box-shadow: -5px 5px 20px 0px rgba(0,0,0,0.05);
    -moz-box-shadow: -5px 5px 20px 0px rgba(0,0,0,0.05);
    box-shadow: -5px 5px 20px 0px rgba(0,0,0,0.05);
  }
}
