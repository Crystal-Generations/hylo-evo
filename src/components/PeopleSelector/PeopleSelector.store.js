import { createSelector } from 'redux-orm'
import { pick } from 'lodash/fp'
import orm from 'store/models'

export const MODULE_NAME = 'PeopleSelector'

export const FETCH_PEOPLE = 'FETCH_PEOPLE'
export const PEOPLE_SELECTOR_SET_AUTOCOMPLETE = 'PEOPLE_SELECTOR_SET_AUTOCOMPLETE'
export const PEOPLE_SELECTOR_ADD_PARTICIPANT = 'PEOPLE_SELECTOR_ADD_PARTICIPANT'
export const PEOPLE_SELECTOR_REMOVE_PARTICIPANT = 'PEOPLE_SELECTOR_REMOVE_PARTICIPANT'

const fetchPeopleQuery =
`query PersonAutocomplete ($autocomplete: String, $first: Int) {
  people (autocomplete: $autocomplete, first: $first) {
    items {
      id
      name
      avatarUrl
      memberships {
        id
        community {
          id
          name
        }
      }
    }
  }
}`

export function fetchPeople (autocomplete, query = fetchPeopleQuery, first = 20) {
  return {
    type: FETCH_PEOPLE,
    graphql: {
      query,
      variables: { autocomplete, first }
    },
    meta: { extractModel: 'Person' }
  }
}

export function addParticipant (id) {
  return {
    type: PEOPLE_SELECTOR_ADD_PARTICIPANT,
    payload: id
  }
}

export function removeParticipant (id) {
  return {
    type: PEOPLE_SELECTOR_REMOVE_PARTICIPANT,
    payload: id
  }
}

export function setAutocomplete (autocomplete) {
  return {
    type: PEOPLE_SELECTOR_SET_AUTOCOMPLETE,
    payload: autocomplete
  }
}

export const participantsSelector = createSelector(
  orm,
  state => state.orm,
  state => state[MODULE_NAME].participants,
  (session, participants) => participants.map(id => pick(
    [ 'id', 'name', 'avatarUrl' ],
    session.Person.withId(id).ref
  ))
)

export const matchesSelector = createSelector(
  orm,
  state => state.orm,
  state => state[MODULE_NAME].autocomplete,
  state => state[MODULE_NAME].participants,
  (session, autocomplete, participants) => {
    if (autocomplete) {
      const term = autocomplete.toLowerCase()
      const matches = session.Person
        .all()
        .filter(p => !participants.includes(p.id))
        .filter(p => p.name.toLowerCase().includes(term))
        .orderBy('name')
      return matches.toModelArray().map(match => ({
        ...pick([ 'id', 'name', 'avatarUrl' ], match.ref),
        community: match.memberships.first()
          ? match.memberships.first().community.name : null
      }))
    }
    return null
  }
)

export const defaultState = {
  participants: []
}

export default function reducer (state = defaultState, action) {
  const { error, payload, type } = action
  if (error) return state

  switch (type) {
    case PEOPLE_SELECTOR_SET_AUTOCOMPLETE:
      return {
        ...state,
        autocomplete: payload
      }

    case PEOPLE_SELECTOR_ADD_PARTICIPANT:
      return {
        ...state,
        participants: [ ...state.participants, payload ]
      }

    case PEOPLE_SELECTOR_REMOVE_PARTICIPANT:
      if (payload) {
        return {
          ...state,
          participants: state.participants.filter(p => p !== payload)
        }
      }
      // No payload? Remove the last element.
      return {
        ...state,
        participants: state.participants.slice(0, state.participants.length - 1)
      }
  }

  return state
}
