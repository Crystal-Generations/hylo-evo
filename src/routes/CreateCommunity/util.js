export const slugValidatorRegex = /^[0-9a-z\s-]+$/
