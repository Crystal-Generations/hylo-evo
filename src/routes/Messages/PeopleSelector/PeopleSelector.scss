.autocomplete-control {
  display: flex;
  flex-wrap: wrap;
  flex: 1;
  align-items: flex-end;
  border-bottom: 1px $color-rhino-30 solid;
  padding-bottom: $space-2x;
  margin-right: $space-5x;
  min-height: $space-13x;
}

.backButton {
  display: none;
}

.autocomplete {
  composes: body-drk-lg from 'css/typography.scss';
  flex: 1;
  border: none;
  outline: none;

  // Yep, believe it or not this is necessary... can't group the selectors!
  &::-webkit-input-placeholder {
    color: $color-rhino-30;
    letter-spacing: 0.4;
  }

  &::-moz-placeholder {
    color: $color-rhino-30;
    letter-spacing: 0.4;
  }

  &:-ms-input-placeholder {
    color: $color-rhino-30;
    letter-spacing: 0.4;
  }

  &:-moz-placeholder {
    color: $color-rhino-30;
    letter-spacing: 0.4;
  }

  &::placeholder {
    color: $color-rhino-30;
    letter-spacing: 0.4;
  }
}

.close-button {
  margin-bottom: $space-4x;
  margin-left: $space-6x;
}

.arrow-down {
  margin-right: 30px;
  margin-top: 10px;
  color: $color-caribbean-green;
  cursor: pointer;
}

.thread-header {
  composes: header from '../Messages.scss';
  padding-top: $space-4x;
  padding-bottom: $space-3x;
  height: unset;
}

@media screen and (max-width: 650px) {
  .backButton {
    display: block;

    span {
      transform: rotate(90deg);
      display: block;
      top: -2px;
    }
  }
}