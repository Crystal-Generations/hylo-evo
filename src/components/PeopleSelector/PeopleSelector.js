import React from 'react'
import { debounce, throttle } from 'lodash/fp'

import Icon from 'components/Icon'
import { getKeyCode, keyMap } from 'util/textInput'
import PeopleSelectorMatches from 'components/PeopleSelectorMatches'
import SelectorMatchedItem from 'components/SelectorMatchedItem'
import './PeopleSelector.scss'

const { any, arrayOf, func, shape, string } = React.PropTypes

const personType = shape({
  id: any,
  name: string,
  avatarUrl: string
})

// TODO: This _grossly_ underestimates the problem! See:
// https://www.w3.org/International/questions/qa-personal-names
const invalidPersonName = /[^a-z '-]+/gi

export default class PeopleSelector extends React.Component {
  static propTypes = {
    autocomplete: string,
    fetchPeople: func,
    deleteParticipant: func,
    participants: arrayOf(personType),
    setAutocomplete: func
  }

  autocompleteSearch = throttle(1000, this.props.fetchPeople)

  onChange = debounce(200, () => {
    const { value } = this.autocomplete
    if (!invalidPersonName.exec(value)) {
      return this.props.setAutocomplete(value)
    }
    this.autocomplete.value = value.replace(invalidPersonName, '')
  })

  onKeyDown (evt) {
    const keyCode = getKeyCode(evt)
    if (keyCode !== keyMap.BACKSPACE) {
      this.autocompleteSearch(this.autocomplete.value)
    }
    if (keyCode === keyMap.COMMA || keyCode === keyMap.ENTER) {
      evt.preventDefault()
      this.autocomplete.value = null
      return this.props.setAutocomplete(null)
    }
  }

  render () {
    const { deleteParticipant, matches, participants } = this.props
    return <div styleName='people-selector'>
      <div styleName='thread-header' tabIndex='0'>
        <div styleName='autocomplete-control'>
          {participants && participants.map(match =>
            <SelectorMatchedItem
              avatarUrl={match.avatarUrl}
              key={match.id}
              name={match.name}
              deleteParticipant={() => deleteParticipant(match.id)} />
          )}
          <input styleName='autocomplete'
            ref={i => this.autocomplete = i} // eslint-disable-line no-return-assign
            type='text'
            spellCheck={false}
            onChange={evt => this.onChange(evt)}
            onKeyDown={evt => this.onKeyDown(evt)}
            placeholder={participants.length ? '' : 'Type in the names of people to message'} />
        </div>
        <Icon name='Ex' styleName='close-button' />
      </div>
      <PeopleSelectorMatches matches={matches} />
    </div>
  }
}
