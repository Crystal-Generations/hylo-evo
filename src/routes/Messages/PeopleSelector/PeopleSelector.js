import PropTypes from 'prop-types'
import React from 'react'
import { debounce, throttle } from 'lodash/fp'
import { getKeyCode, keyMap } from 'util/textInput'
import CloseMessages from '../CloseMessages'
import PeopleSelectorMatches from './PeopleSelectorMatches'
import PeopleSelectorContacts from './PeopleSelectorContacts'
import SelectorMatchedItem from './SelectorMatchedItem'
import './PeopleSelector.scss'

const invalidPersonName = /[^a-z '-]+/gi

export default class PeopleSelector extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      currentMatch: null
    }
    this.autocompleteInput = React.createRef()
  }

  componentDidMount () {
    // TODO: Is there a better way to handle empty functions coming-back from Apollo
    this.props.fetchDefaultList && this.props.fetchDefaultList()
  }

  componentWillReceiveProps (props) {
    const { matches } = props
    if (!matches || matches.length === 0) {
      this.setCurrentMatch()
      return
    }

    if (matches.find(m => this.state.currentMatch && m.id === this.state.currentMatch.id)) return
    this.setState({ currentMatch: matches[0] })
  }

  addParticipant = (participant) => {
    this.autocompleteInput.current.focus()
    if (this.props.participants.find(p => p.id === participant.id)) return
    this.props.setContactsSearch(null)
    this.setCurrentMatch()
    this.autocompleteInput.current.value = null
    this.props.addParticipant(participant)
  }

  removeParticipant = (participant) => {
    this.props.removeParticipant(participant)
  }

  excludeParticipantsAndCurrentUser = (contactList) => {
    if (!contactList) return
    const { currentUser, participants } = this.props
    const participantsIds = participants.map(p => p.id)
    return contactList
      .filter(c => currentUser ? c.id !== currentUser.id : true)
      .filter(c => !participantsIds.includes(c.id))
  }

  arrow (direction, event) {
    event.preventDefault()
    let delta = 0
    const idx = this.props.matches.findIndex(m => m.id === this.state.currentMatch.id)
    if (direction === 'up') {
      if (idx > 0) delta = -1
    }
    if (direction === 'down') {
      if (idx < this.props.matches.length - 1) delta = 1
    }
    this.setCurrentMatch(this.props.matches[idx + delta])
  }

  autocompleteSearch = throttle(1000, this.props.fetchPeople)

  onChange = debounce(200, () => {
    const { value } = this.autocompleteInput.current
    if (!invalidPersonName.exec(value)) {
      return this.props.setContactsSearch(value)
    }
    this.autocompleteInput.current.value = value.replace(invalidPersonName, '')
  })

  onKeyDown (evt) {
    switch (getKeyCode(evt)) {
      case keyMap.BACKSPACE: return this.state.currentMatch ? null : this.removeParticipant()
      case keyMap.UP: return this.arrow('up', evt)
      case keyMap.DOWN: return this.arrow('down', evt)
      case keyMap.COMMA:
      case keyMap.ENTER:
        evt.preventDefault()
        this.addParticipant(this.state.currentMatch)
        this.autocompleteInput.current.value = null
        return this.props.setContactsSearch(null)

      default:
        this.autocompleteSearch(this.autocompleteInput.current.value)
    }
  }

  setCurrentMatch = person => this.setState(() => ({ currentMatch: person }))

  render () {
    const {
      contacts,
      recentContacts,
      matches,
      onCloseURL,
      participants
    } = this.props
    const {
      currentMatch
    } = this.state

    return <React.Fragment>
      <div styleName='thread-header' tabIndex='0'>
        <div styleName='autocomplete-control'>
          {participants && participants.map(participant =>
            <SelectorMatchedItem
              avatarUrl={participant.avatarUrl}
              key={participant.id}
              name={participant.name}
              removeParticipant={() => this.removeParticipant(participant)} />
          )}
          <input styleName='autocomplete'
            autoFocus
            ref={this.autocompleteInput}
            type='text'
            spellCheck={false}
            onChange={evt => this.onChange(evt)}
            onKeyDown={evt => this.onKeyDown(evt)}
            placeholder={participants.length ? '' : 'Type in the names of people to message'} />
        </div>
        <CloseMessages onCloseURL={onCloseURL} />
      </div>
      {currentMatch
        ? <PeopleSelectorMatches
          addParticipant={this.addParticipant}
          currentMatch={currentMatch}
          matches={this.excludeParticipantsAndCurrentUser(matches)}
          setCurrentMatch={this.setCurrentMatch} />
        : <PeopleSelectorContacts
          addParticipant={this.addParticipant}
          contacts={this.excludeParticipantsAndCurrentUser(contacts)}
          recentContacts={this.excludeParticipantsAndCurrentUser(recentContacts)} />}
    </React.Fragment>
  }
}

PeopleSelector.propTypes = {
  addParticipant: PropTypes.func.isRequired,
  contacts: PropTypes.array,
  currentUser: PropTypes.object,
  fetchPeople: PropTypes.func,
  fetchDefaultList: PropTypes.func,
  matches: PropTypes.array,
  onCloseURL: PropTypes.string,
  participants: PropTypes.array,
  recentContacts: PropTypes.array,
  removeParticipant: PropTypes.func,
  setContactsSearch: PropTypes.func
}
