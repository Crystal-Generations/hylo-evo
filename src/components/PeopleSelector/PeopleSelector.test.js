import { mount, shallow } from 'enzyme'
import React from 'react'

import { keyMap } from 'util/textInput'
import PeopleSelector from './PeopleSelector'


it('matches the last snapshot', () => {
  const matches = [
    { id: '1', name: 'Wombat' },
    { id: '2', name: 'Aardvark' }
  ]
  const wrapper = shallow(
    <PeopleSelector
      deleteMatch={jest.fn()} 
      fetchPeople={jest.fn()}
      matches={matches} />
  )
  expect(wrapper).toMatchSnapshot()
})

it('does not hit server when backspace is pressed', () => {
  const fetchPeople = jest.fn()
  const wrapper = mount(
    <PeopleSelector matches={[]} fetchPeople={fetchPeople} />
  )
  wrapper.find('input').first().simulate('keyDown', { keyCode: keyMap.BACKSPACE })
  expect(fetchPeople).not.toHaveBeenCalled()
})

//it('hits server when keys other than backspace are pressed', () => {
  //const fetchPeople = jest.fn()
  //const wrapper = mount(
    //<PeopleSelector matches={[]} fetchPeople={fetchPeople} setAutocomplete={() => {}} />
  //)
  //wrapper.find('input').first().simulate('keyDown', { keyCode: keyMap.SPACE })
  //expect(fetchPeople).toHaveBeenCalled()
//})

//describe('setAutocomplete', () => {
  //beforeEach(() => {
    //jest.useFakeTimers()
  //})

  //it('updates if user input contains valid characters', () => {
    //const expected = 'Poor Yorick'
    //const setAutocomplete = jest.fn()
    //const wrapper = mount(
      //<PeopleSelector fetchPeople={() => {}} matches={[]} setAutocomplete={setAutocomplete} />
    //)
    //const input = wrapper.find('input').first()
    //input.node.value = expected
    //input.simulate('change')
    //jest.runAllTimers()

    //const actual = setAutocomplete.mock.calls[0][0]
    //expect(actual).toBe(expected)
  //})

  //it('does not update if user input contains invalid characters', () => {
    //const invalid = 'Poor Yorick9238183$@#$$@!'
    //const expected = 'Poor Yorick'
    //const setAutocomplete = jest.fn()
    //const wrapper = mount(
      //<PeopleSelector fetchPeople={() => {}} matches={[]} setAutocomplete={setAutocomplete} />
    //)
    //const input = wrapper.find('input').first()
    //input.node.value = invalid
    //input.simulate('change')

    //jest.runAllTimers()
    //expect(setAutocomplete).not.toHaveBeenCalled()
    //expect(input.node.value).toBe(expected)
  //})
//})
