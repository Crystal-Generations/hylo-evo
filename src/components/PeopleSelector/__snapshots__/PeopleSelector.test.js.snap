// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the last snapshot 1`] = `
<div
  data-styleName="people-selector"
>
  <div
    data-styleName="thread-header"
    tabIndex="0"
  >
    <PeopleSelector />
    <Icon
      data-styleName="close-button"
      name="Ex"
    />
  </div>
  <div
    data-styleName="autocomplete-control"
  >
    <SelectorMatchedItem
      deleteMatch={[Function]}
      name="Wombat"
    />
    <SelectorMatchedItem
      deleteMatch={[Function]}
      name="Aardvark"
    />
    <input
      data-styleName="autocomplete"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder=""
      spellCheck={false}
      type="text"
    />
  </div>
</div>
`;
